agent/csi/plugin.go

```
@@ -33,12 +50,133 @@ type NodePlugin struct {
+		if volInfo.isPublished {
+			return volInfo.targetPath, nil
+		}
+		return volInfo.targetPath, fmt.Errorf("volume %s is not published", volumeID)
+	}
+	return "", nil
 }
 
 func (np *NodePlugin) NodeGetInfo(ctx context.Context) (*api.NodeCSIInfo, error) {
+	np.Lock.RLock()
+	defer np.Lock.RUnlock()
 	resp := &csi.NodeGetInfoResponse{
 		NodeId: np.NodeID,
 	}
 
 	return makeNodeInfo(resp), nil
 }
+
+func (np *NodePlugin) NodeStageVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+	stagingTarget := fmt.Sprintf(TargetPath, volID)
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.Lock.Lock()
+	defer np.Lock.Unlock()
+
+	v := &volumePublishStatus{
+		targetPath: stagingTarget,
+	}
+
+	np.VolumeMap[volID] = v
+
+	return nil
+}
+
+func (np *NodePlugin) NodeUnstageVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.Lock.Lock()
+	defer np.Lock.Unlock()
+	if v, ok := np.VolumeMap[volID]; ok {
+		if v.isPublished {
+			return status.Errorf(codes.InvalidArgument, "VolumeID %s is not unpublished", volID)
+		}
+		delete(np.VolumeMap, volID)
+		return nil
+	}
+
+	return status.Errorf(codes.NotFound, "VolumeID %s is not staged", volID)
+}
+
+func (np *NodePlugin) NodePublishVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.Lock.Lock()
+	defer np.Lock.Unlock()
+	if v, ok := np.VolumeMap[volID]; ok {
+		if _, err := os.Stat(v.targetPath); os.IsNotExist(err) {
+			os.Mkdir(v.targetPath, os.ModeDir)
+		}
+		v.isPublished = true
+		return nil
+	}
+
+	return status.Errorf(codes.NotFound, "VolumeID %s is not staged", volID)
```

Two things:

1. This error should likely be whatever is returned from the CSI plugin
2. Since we're not actually calling the CSI plugin yet, this should return the same error one would expect from the CSI plugin. Trying to publish a volume that is not staged returns `FAILED_PRECONDITION`.